src/controlplane/client/connector.go

```diff
@@ -0,0 +1,230 @@
+package client
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/newrelic/infra-integrations-sdk/log"
+	"github.com/newrelic/nri-kubernetes/v2/internal/config"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/transport"
+
+	"github.com/newrelic/nri-kubernetes/v2/src/client"
+)
+
+const (
+	DefaultTimout      = 5000 * time.Millisecond
+	defaultMetricsPath = "/metrics"
+	mTLSAuth           = "mTLS"
+	bearerAuth         = "bearer"
+)
+
+// Connector provides an interface to retrieve []connParams to connect to a Control Plane instance.
+type Connector interface {
+	Connect() (*connParams, error)
+}
+
+type defaultConnector struct {
+	// TODO: Use a non-sdk logger
+	logger          log.Logger
+	kc              kubernetes.Interface
+	inClusterConfig *rest.Config
+	endpoints       []config.Endpoint
+}
+
+// DefaultConnector returns a defaultConnector that probes all endpoints in the list and return the first responding status OK.
+func DefaultConnector(endpoints []config.Endpoint, kc kubernetes.Interface, inClusterConfig *rest.Config, logger log.Logger) (Connector, error) {
+	return &defaultConnector{
+		logger:          logger,
+		inClusterConfig: inClusterConfig,
+		kc:              kc,
+		endpoints:       endpoints,
+	}, nil
+}
+
+// Connect iterates over the endpoints list probing each endpoint with a HEAD request
+// and returns the connection parameters of the first endpoint that respond Status OK.
+func (dp *defaultConnector) Connect() (*connParams, error) {
+	for _, e := range dp.endpoints {
+		dp.logger.Debugf("Configuring endpoint %q for probing", e.URL)
+
+		u, err := url.Parse(e.URL)
+		if err != nil {
+			return nil, fmt.Errorf("parsing endpoint url %q: %w", e.URL, err)
+		}
+
+		if strings.TrimSuffix(u.Path, "/") == "" {
+			dp.logger.Debugf("Autodiscover endpoint %q does not contain path, adding default %q", e.URL, defaultMetricsPath)
+			u.Path = defaultMetricsPath
+		}
+
+		httpClient, err := dp.newHTTPClient(e)
+		if err != nil {
+			return nil, fmt.Errorf("creating HTTP client for endpoint %q: %w", e.URL, err)
+		}
+
+		if err := dp.probeEndpoint(u.String(), httpClient); err != nil {
+			dp.logger.Debugf("Endpoint %q probe failed, skipping: %v", e.URL, err)
+			continue
+		}
+
+		dp.logger.Debugf("Endpoint %q probed successfully", e.URL)
```

Comment:
I'd bump this to `Infof` if we're not logging the chosen endpoint anywhere else, it could be useful information.

Reply:
the probing happen on each run of the scraper, wouldn't be this a bit verbose?